include/ensmallen_bits/nsga2/nsga2.hpp

```diff
@@ -0,0 +1,274 @@
+ * see the documentation on function types included with this distribution or
+ * on the ensmallen website.
+ */
+class NSGA2 {
+ public:
+  /**
+   * Constructor for the NSGA2 optimizer
+   *
+   * The default values provided over here are not necessarily suitable for a
+   * given function. Therefore it is highly recommended to adjust the
+   * parameters according to the problem.
+   *
+   * @param populationSize The number of candidates in the population.
+   *     This should be atleast 4 in size and a multiple of 4.
+   * @param maxGenerations The maximum number of generations allowed for NSGA-II
+   * @param crossoverProb The probability that a crossover will occur.
+   * @param mutationProb The probability that a mutation will occur.
+   * @param mutationStrength The strength of the mutation.
+   * @param epsilon The minimum difference reqiured to distinguish between
+   *     candidate solutions.
+   */
+  NSGA2(const size_t populationSize = 100,
+        const size_t maxGenerations = 2000,
+        const double crossoverProb = 0.6,
+        const double mutationProb = 0.3,
+        const double mutationStrength = 1e-3,
+        const double epsilon = 1e-6);
+
+  /**
+   * Optimize a set of objectives. The initial population is generated using the
+   * starting point. The output is the best generated front.
+   *
+   * @tparam MultiobjectiveFunctionType A generic multi-objective function.
+   * @tparam MatType Type of matrix to optimize.
+   * @tparam CallbackTypes Types of callback functions.
+   * @param objectives Vector of objective functions to optimize for.
+   * @param iterate Starting point
+   * @param callbacks Callback functions.
+   * @return std::vector<MatType> The best front obtained after optimzation.
+   */
+  template<typename MultiobjectiveFunctionType,
+           typename MatType,
+           typename... CallbackTypes>
+  std::vector<MatType> Optimize(MultiobjectiveFunctionType& objectives,
+                                MatType& iterate,
+                                CallbackTypes&&... callbacks);
+
+  //! Get the population size.
+  size_t PopulationSize() const { return populationSize; }
+  //! Modify the population size.
+  size_t& PopulationSize() { return populationSize; }
+
+  //! Get the maximum number of generations.
+  size_t MaxGenerations() const { return maxGenerations; }
+  //! Modify the maximum number of generations.
+  size_t& MaxGenerations() { return maxGenerations; }
+
+  //! Get the crossover rate.
+  double CrossoverRate() const { return crossoverProb; }
+  //! Modify the crossover rate.
+  double& CrossoverRate() { return crossoverProb; }
+
+  //! Get the mutation probability.
+  double MutationProbability() const { return mutationProb; }
+  //! Modify the mutation probability.
+  double& MutationProbability() { return mutationProb; }
+
+  //! Get the mutation size.
+  double MutationSize() const { return mutationStrength; }
```

Comment:
Should we rename the function to `MutationStrength()` as this is what the parameter is called?

Reply:
@favre49 suggested that we rename it to `MutationSize` as CNE uses the same name. But, CNE has the variable named `mutationStrength` too, so that makes sense for CNE. We can rename it MutationSize, what do you say?